monasca: Add Apache Cassandra support (SCRD-6731)

witekest opened the pull request:
The change adds the option to install Apache Cassandra as the TSDB.
Currently only single node installation is supported.
Choosing this option implicitly installs Java implementation of
monasca-persister.

jgrassler commented on chef/data_bags/crowbar/template-monasca.json:
```diff
@@ -124,15 +124,16 @@
         "keystone_monasca_operator_password": "",
         "keystone_monasca_agent_password": "",
         "keystone_admin_agent_password": "",
-        "tsdb": "influxdb",
```
Just a question: why did you move this away from master? I'm ok with it, but it _is_ going to require a documentation update :-)

toabctl replied:
There is no `master` role anymore. I think we should try to remove all remaining options from master

witekest replied:
+1

witekest replied:
> Just a question: why did you move this away from master? I'm ok with it, but it _is_ going to require a documentation update :-)

Setting passwords is [not documented](http://docserv.nue.suse.com/documents/SUSE_OpenStack_Cloud_9/suse-openstack-cloud-deployment/single-html/#id-1.3.5.4.25.18).

toabctl commented on chef/cookbooks/monasca/recipes/cassandra.rb:
```diff
@@ -0,0 +1,75 @@
+
+package "cassandra"
+
+monasca_servers = search(:node, "roles:monasca-server")
```
there is `node_search_with_cache` . any reason why tho not use that? In envs with a lot of nodes, `node_search_with_cache` is prefered.

witekest replied:
I copied it from your code in InfluxDB recipe :)
The point is valid and I fixed it in both recipes.

toabctl commented on chef/cookbooks/monasca/attributes/default.rb:
```diff
@@ -108,6 +108,18 @@
+default[:monasca][:cassandra][:batch_size_warn_threshold_in_kb] = 2000
+default[:monasca][:cassandra][:batch_size_fail_threshold_in_kb] = 2500
+default[:monasca][:cassandra][:admin_user] = "cassandra"
+default[:monasca][:cassandra][:admin_default_password] = "cassandra"
```
is that the password that is used when the package is freshly installed?

witekest replied:
Yes.

toabctl commented on chef/cookbooks/monasca/recipes/cassandra.rb:
```diff
@@ -0,0 +1,75 @@
+  block do
+    CassandraHelper.set_password(admin_role, admin_password,
+                                 user: admin_user,
+                                 password: default_password,
```
shouldn't that be the `admin_password` ?

witekest replied:
No, we log in with `default_password` and set it to `admin_password`.

toabctl commented on chef/cookbooks/monasca/recipes/monasca_api.rb:
```diff
@@ -68,7 +70,7 @@
     node.save
   end
-  not_if { node[:monasca][:db_monapi_synced] }
+  not_if {node[:monasca][:db_monapi_synced]}
```
change not needed

witekest replied:
Reverted.

toabctl commented on chef/cookbooks/monasca/recipes/monasca_persister_java.rb:
```diff
@@ -0,0 +1,40 @@
+
+package "openstack-monasca-persister-java"
+
+monasca_servers = search(:node, "roles:monasca-server")
```
`node_search_with_cache` imo

witekest replied:
Fixed both in _monasca_perister_ and _monasca_persister_java_ recipes.

toabctl commented on chef/cookbooks/monasca/recipes/monasca_persister_java.rb:
```diff
@@ -0,0 +1,40 @@
+  group node[:monasca][:persister][:group]
+  mode "0640"
+  variables(
+    zookeeper_host: monasca_net_ip,
```
where is `zookeeper_host` defined?

witekest replied:
I don't understand the question. `zookeeper_host` is the variable used in _monasca-persister-java.yaml.erb_ template.

toabctl replied:
urgs. not enough tee in the morning. sorry

toabctl commented on chef/cookbooks/monasca/recipes/monasca_persister_java.rb:
```diff
@@ -0,0 +1,40 @@
+  mode "0640"
+  variables(
+    zookeeper_host: monasca_net_ip,
+    cassandra_hosts: monasca_net_ip
```
same here

toabctl commented:
@witekest I tested this manually and it failed with:
```
FATAL: RuntimeError: ruby_block[Create role for monasca_api] (monasca::cassandra line 68) had an error: RuntimeError: Can not execute '/usr/bin/cqlsh --user cassandra --password gluCm3hXXwxp --execute "CREATE ROLE IF NOT EXISTS monasca_api WITH PASSWORD='gluCm3hXXwxp' AND LOGIN = true" 192.168.192.82': Connection error: ('Unable to connect to any servers', {'192.168.192.82': error(111, "Tried connecting to [('192.168.192.82', 9042)]. Last error: Connection refused")})
```

witekest commented:
> @witekest I tested this manually and it failed with:
> 
> ```
> FATAL: RuntimeError: ruby_block[Create role for monasca_api] (monasca::cassandra line 68) had an error: RuntimeError: Can not execute '/usr/bin/cqlsh --user cassandra --password gluCm3hXXwxp --execute "CREATE ROLE IF NOT EXISTS monasca_api WITH PASSWORD='gluCm3hXXwxp' AND LOGIN = true" 192.168.192.82': Connection error: ('Unable to connect to any servers', {'192.168.192.82': error(111, "Tried connecting to [('192.168.192.82', 9042)]. Last error: Connection refused")})
> ```
> 
> You might need to wait a bit or retry after cassandra starts. There is no SD_NOTIFY support in cassandra

Thanks for catching. I have moved this section to _monasca-api_ recipe, so Cassandra should have enough time to start. I have also added `retries` in the following block, as I got AuthenticationError when testing in the same environment.

toabctl commented:
Triggered a CI build with monasca/cassandra
